src/caffe/layers/batch_norm_layer.cu

```diff
@@ -6,166 +6,205 @@
 
 namespace caffe {
 
+// multicast x[c] into y[.,c,...]
+template <typename Dtype>
+void BatchNormLayer<Dtype>::multicast_gpu(int N, int C, int S,
+    const Dtype *x, Dtype *y ) {
+  Blob<Dtype> temp_NC;
+  vector<int> temp_size;
+  temp_size.push_back(N*C);
+  temp_NC.Reshape(temp_size);
+  caffe_gpu_gemm<Dtype>(CblasNoTrans, CblasNoTrans, N, C, 1,
+      1., ones_N_.gpu_data(), x, 0., temp_NC.mutable_gpu_data());
+  caffe_gpu_gemm<Dtype>(CblasNoTrans, CblasNoTrans, N*C, S , 1,
+      1., temp_NC.gpu_data(), ones_HW_.gpu_data(), 0., y);
+}
+
+// y[c] = sum x(.,c,...)
+template <typename Dtype>
+void BatchNormLayer<Dtype>::compute_sum_per_channel_gpu(int N, int C, int S,
+    const Dtype *x, Dtype *y ) {
+  // assume that x.shape(1)==C
+  Blob<Dtype> temp_NC;
```

Comment:
Originally it was temp_NC_, but then I decided that it can be more safe to allocate temp_NC inside of these 2 functions to avoid potential hidden over-write (for example if I will use temp_NC_ outside) . I will re-check if this have noticable effect on time.